pyro/contrib/epidemiology/util.py

```diff
@@ -2,14 +2,82 @@
 # SPDX-License-Identifier: Apache-2.0
 
 import numpy
-
 import torch
 
 import pyro
 import pyro.distributions as dist
+import pyro.poutine as poutine
+from pyro.distributions.util import broadcast_shape
 from pyro.ops.tensor_utils import safe_log
 
 
+def clamp(tensor, *, min=None, max=None):
+    """
+    Like :func:`torch.clamp` but dispatches to :func:`torch.min` and/or
+    :func:`torch.max` if ``min`` and/or ``max`` is a :class:`~torch.Tensor`.
+    """
+    if isinstance(min, torch.Tensor):
+        tensor = torch.max(tensor, min)
+        min = None
+    if isinstance(max, torch.Tensor):
+        tensor = torch.min(tensor, max)
+        max = None
+    if min is None and max is None:
+        return tensor
+    return tensor.clamp(min=min, max=max)
+
+
+def cat2(lhs, rhs, *, dim=-1):
+    """
+    Like ``torch.cat([lhs, rhs], dim=dim)`` but dispatches to
+    :func:`torch.nn.functional.pad` in case one of ``lhs`` or ``rhs`` is a
+    scalar.
+    """
+    assert dim < 0
+    if not isinstance(lhs, torch.Tensor):
+        pad = (0, 0) * (-1 - dim) + (1, 0)
+        return torch.nn.functional.pad(rhs, pad, value=lhs)
+    if not isinstance(rhs, torch.Tensor):
+        pad = (0, 0) * (-1 - dim) + (0, 1)
+        return torch.nn.functional.pad(lhs, pad, value=rhs)
+
+    diff = lhs.dim() - rhs.dim()
+    if diff > 0:
+        rhs = rhs.expand((1,) * diff + rhs.shape)
+    elif diff < 0:
+        diff = -diff
+        lhs = lhs.expand((1,) * diff + lhs.shape)
+    shape = list(broadcast_shape(lhs.shape, rhs.shape))
+    shape[dim] = -1
+    return torch.cat([lhs.expand(shape), rhs.expand(shape)], dim=dim)
+
+
+@torch.no_grad()
+def align_samples(samples, model, particle_dim):
+    """
+    Unsqueeze stacked samples such that their particle dim all aligns.
+    This traces ``model`` to determine the ``event_dim`` of each site.
+    """
+    assert particle_dim < 0
+
+    sample = {name: value[0] for name, value in samples.items()}
+    with poutine.block(), poutine.trace() as tr, poutine.condition(data=sample):
+        model()
+
+    samples = samples.copy()
+    for name, value in samples.items():
+        event_dim = tr.trace.nodes[name]["fn"].event_dim
+        pad = event_dim - particle_dim - value.dim()
+        if pad < 0:
+            raise ValueError("Cannot align samples, try moving particle_dim left")
+        if pad > 0:
+            shape = value.shape[:1] + (1,) * pad + value.shape[1:]
+            print("DEBUG reshaping {} : {} -> {}".format(name, value.shape, shape))
+            samples[name] = value.reshape(shape)
+
+    return samples
```

Comment:
@neerajprad I believe we can later refactor to make this generic, say as an `align_samples` or `particle_dim` kwarg to `MCMC.get_samples()`.

Reply:
That sounds reasonable. IIUC, we can put an outermost particle dim with size=1 for a model (if `align_samples=True`), collect samples with our usual flow and concatenate the collected samples at dim 0 which should be the same as sampling from a vectorized model with an outermost plate dim. Would that work?

Reply:
I think it's safer to first collect samples and `event_dim` s and afterwards reshape. That way we wouldn't force users to write vectorizable / broadcastable code.

Reply:
Sorry, I missed your comment. That sounds reasonable, but would would sample reshaping be useful if the model wasn't vectorizable? Just want to make sure that I understand the use case.

Reply:
Yes sample reshaping would still be useful even if the model is not vectorizable. For example in the `CompartmentalModel` in this PR, there are three models that are mathematically equivalent but with different computational complexity. We run HMC on one of those modes that is not vectorizable over particles (it is instead vectorized over time, hence its name `_vectorized_model()`). We then stitch together multiple samples and `poutine.condition` two other model that are vectorized over particles but are sequential over time (`_sequential_model()` and `_generative_model()`).

Reply:
Thanks for explaining, this makes sense. I'll go over the models and your reshaping utility.